File: ai/bot/helper/count_player.py
##
## EPITECH PROJECT, 2025
## B-YEP-400-STG-4-1-zappy-noe.carabin
## File description:
## count_player
##

from ... import ai_print as ap
from ... import ai_threads as at

def count_players_on_tile(look_string, tile_index, thread_name="UnknownThread"):
    """Counts players on a specific tile from look results. Returns -1 on error."""

    if not look_string or not look_string.startswith('[') or not look_string.endswith(']'):
        ap.safe_print(f"[{thread_name}] CRITICAL PARSE ERROR: Bad Look string for player count: '{look_string}'")
        at.trigger_exit(f"{thread_name} PlayerCountParseFormat")
        return -1

    try:
        tiles = look_string.strip('[]').split(',')
        if 0 <= tile_index < len(tiles):
            return tiles[tile_index].split().count('player')
        ap.safe_print(f"[{thread_name}] WARN: Tile index {tile_index} out of bounds for look string.")
        return 0
    except Exception as e:
        ap.safe_print(f"[{thread_name}] CRITICAL PARSE ERROR: Failed counting players in look string '{look_string}': {e}")
        at.trigger_exit(f"{thread_name} PlayerCountException")
        return -1

File: ai/entity/bot/helper/count_player.cpp
/*
** EPITECH PROJECT, 2025
** B-YEP-400-STG-4-1-zappy-noe.carabin
** File description:
** count_player
*/

#include "../../AI.hpp"
#include "../../../utils/utils.hpp"
#include <algorithm>

int ai::entity::AI::countPlayersOnTile(int idx, const std::string &look)
{
    utils::debug::Logger &logger = utils::debug::Logger::GetInstance();

    if (look.empty() || look.front() != '[' || look.back() != ']') {
        logger.log("[Error] Bad look format.");
        return -1;
    }

    const std::string content = look.substr(1, look.size() - 2);
    const std::vector<std::string> tiles = utils::string::split(content, ',');

    if (idx >= 0 && idx < static_cast<int>(tiles.size())) {
        const std::string tile = tiles[idx];
        const std::vector<std::string> words = utils::string::split(tile, ' ');
        return std::count(words.begin(), words.end(), "player");
    }
    logger.log("Tile index " + std::to_string(idx) + " out of bounds for look string.");
    return 0;
}
